Fix: prevent memory blowout in LoggingWorker

AlexsanderHamir opened the pull request:
## Title

Fix: prevent memory blowout in LoggingWorker

## Relevant issues

Addresses #12685

<!-- e.g. "Fixes #000" -->

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [ ] I have added a screenshot of my new test passing locally 
- [x] My PR passes all unit tests on [`make test-unit`](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] My PR's scope is as isolated as possible, it only solves 1 specific problem


## Type

<!-- Select the type of Pull Request -->
<!-- Keep only the necessary ones -->

🐛 Bug Fix
🧹 Refactoring

## Memory Impact

**Before**
<img width="2696" height="994" alt="image" src="https://github.com/user-attachments/assets/b60cfc86-1c76-4707-b15f-87ce5d62e2e0" />


**After**
<img width="2722" height="1050" alt="image" src="https://github.com/user-attachments/assets/44dfac40-09cc-4b10-a074-075ea33fffa5" />


## Context

Tasks were being executed sequentially with each task awaited before processing the next one. When the queue had 10k+ tasks, only one could execute at a time. Since the request rate exceeded execution speed, objects accumulated in memory (50k+), holding references to heavy objects and causing memory blowout.

The new implementation uses a semaphore to allow up to 100 concurrent tasks while properly tracking and cleaning up each task, significantly improving throughput and preventing queue buildup.

## Changes

- Added `concurrency` parameter to `__init__` with default value of 100
- Added `self._sem` semaphore to limit concurrent task execution
- Added `self._running_tasks` set to track spawned tasks
- Initialized semaphore in `start()` method
- Replaced sequential `await` in `_worker_loop` with `asyncio.create_task()` to spawn concurrent tasks
- Added task tracking via `add_done_callback()` for automatic cleanup
- Created new `_process_log_task()` method to handle individual task execution with semaphore
- Updated `stop()` method to cancel and await all running tasks before cleanup
- Added early return guards in `_worker_loop()` and `_process_log_task()` when queue/semaphore not initialized
- Made logger constants configurable
- Implemented configurable cleaning of the queue once its full to avoid dropping tasks

## Notes

- Fixes critical issue where LiteLLM pods would crash within 15 minutes under 1k concurrent users due to OOM. However, memory issues remain to be addressed, including allocation patterns that prevent the Python allocator from efficiently releasing memory, causing fragmentation and accumulation.

- The concurrency is limited to 100 as a tradeoff: executing tasks sequentially (old approach) blocked everything and caused queue buildup, while unrestricted concurrency would starve incoming requests. The semaphore balances throughput with system stability.

AlexsanderHamir committed:
fix: prevent memory blowout in LoggingWorker

Tasks were being executed sequentially with each task awaited before
processing the next one. When the queue had 10k+ tasks, only one could
execute at a time. Since the request rate exceeded execution speed,
objects accumulated in memory (50k+), holding references to heavy
objects and causing memory blowout.

The new implementation uses a semaphore to allow up to 1000 concurrent
tasks while properly tracking and cleaning up each task, significantly
improving throughput and preventing queue buildup.

ishaan-jaff reviewed (changes requested):
does this have any impact on RPS or latency ?

AlexsanderHamir commented:
> does this have any impact on RPS or latency ?

There was not negative impact, it actually improved p95.

An important caveat is that ANY change to `LOGGING_WORKER_CONCURRENCY` should be tested to ensure no performance regression was caused, I found 1k to be a good number where the worker is no longer drowning but we aren't taking too much space on the event loop.

AlexsanderHamir committed:
fix: require semaphor before removing task from queue

AlexsanderHamir committed:
fix: make worker concurrency configurable

AlexsanderHamir committed:
fix: clean comments

AlexsanderHamir committed:
Merge branch 'main' into litellm_memory_issues_05

AlexsanderHamir committed:
fix: clarify new env purpose

AlexsanderHamir committed:
Merge branch 'litellm_memory_issues_05' of https://github.com/BerriAI/litellm into litellm_memory_issues_05

AlexsanderHamir committed:
fix: add missing lib

AlexsanderHamir committed:
make constants configurable instead of hardcoded

AlexsanderHamir committed:
add more aggressive cleaning when queue is full

AlexsanderHamir committed:
add helpers function for the aggressive cleaning functionality

AlexsanderHamir committed:
use envs instead of static constants

AlexsanderHamir committed:
import and document constants

AlexsanderHamir committed:
add unit test for new functionality

AlexsanderHamir committed:
fix default value on config_settings

AlexsanderHamir committed:
Merge remote-tracking branch 'origin/main' into litellm_memory_issues_05

AlexsanderHamir committed:
Merge remote-tracking branch 'origin/main' into litellm_memory_issues_05

AlexsanderHamir committed:
fix: remove unused variables and imports to resolve linter errors

- Remove unused time_since_last_clear variable in logging_worker.py
  The variable was calculated but never used in _handle_queue_full()
  method, causing F841 linter error.

- Remove unused TYPE_CHECKING import in mcp_server/server.py
  The import was not used anywhere in the file, causing F401 linter error.

These changes improve code cleanliness and ensure the codebase passes
all linter checks without affecting functionality.

AlexsanderHamir committed:
add missing log expected by test_queue_full_handling

AlexsanderHamir committed:
fix: clean config_setting.md file

AlexsanderHamir committed:
fix: handle logging errors gracefully during shutdown in _flush_on_exit

During process shutdown, logging handlers may be closed while _flush_on_exit
tries to flush queued logging coroutines. This causes 'ValueError: I/O
operation on closed file' errors when coroutines attempt to log.

Changes:
- Add _safe_log helper method that wraps logging calls and suppresses
  errors when logging handlers are closed (ValueError, OSError, AttributeError)
- Replace all verbose_logger calls in _flush_on_exit with _safe_log
- Remove logging from exception handler in coroutine execution loop
  to prevent cascading errors during shutdown

This ensures graceful shutdown even when logging handlers are closed,
which is common during process termination.

ishaan-jaff reviewed (approved):
lgtm
